docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,304 @@
+requires less direct intervention from infrastructure providers and puts more resposibility on users, for
+this additional responsibility the users gain immediate access to the feature. These methods are
+mutually exclusive, and the annotations will take preference when specified.
+
+#### Infrastructure Machine Template Status Updates
+
+Infrastructure providers should add a field to the `status` of any Infrastructure Machine Template they reconcile.
+This field will contain the CPU, memory, and GPU resources associated with the machine described by
+the template.  Internally, this field will be represented by a Go `map` type utilizing named constants
+for the keys and `k8s.io/apimachinery/pkg/api/resource.Quantity` as the values (similar to how resource
+limits and requests are handled for pods).
+
+It is worth mentioning that the Infrastructure Machine Templates are not usually reconciled by themselves.
+Each infrastructure provider will be responsible for determining the best implementation for adding the
+status field based on the information available on their platform.
+
+**Example implementation in Docker provider**
+```
+// these constants will be carried in Cluster API, but are repeated here for clarity
+const (
+    AutoscalerResourceCPU corev1.ResourceName = "cpu"
+    AutoscalerResourceMemory corev1.ResourceName = "memory"
+    AutoscalerResourceGPU corev1.ResourceName = "gpu"
+)
+
+// AutoscalerMachineStatus defines information about individual machines used by the cluster autoscaler to perform scale from zero operations
+type AutoscalerMachineStatus struct {
+    Capacity corev1.ResourceList `json:"capacity,omitempty"`
+}
+
+// AutoscalerMachineTemplateStatus defines information that is used by the cluster autoscaler to perform scale from zero operations
+type AutoscalerMachineTemplateStatus struct {
+    Machine AutoscalerMachineStatus `json:"machine,omitempty"`
+}
+
+// DockerMachineTemplateStatus defines the observed state of a DockerMachineTemplate
+type DockerMachineTemplateStatus struct {
+    Autoscaler AutoscalerMachineTemplateStatus `json:"autoscaler,omitempty"`
+}
+
+// DockerMachineTemplate is the Schema for the dockermachinetemplates API.
+type DockerMachineTemplate struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec DockerMachineTemplateSpec     `json:"spec,omitempty"`
+    Status DockerMachineTemplateStatus `json:"status,omitempty"`
+}
+```
+_Note: the `ResourceList` and `ResourceName` referenced are from k8s.io/api/core/v1`_
+
+When used as a manifest, it would look like this:
+
+```
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: DockerMachineTemplate
+metadata:
+  name: workload-md-0
+  namespace: default
+spec:
+  template:
+    spec: {}
+status:
+  autoscaler:
+    machine:
+      capacity:
+        memory: 500mb
+        cpu: "1"
+        gpu: "1"
+```
+
+#### MachineSet and MachineDeployment Annotations
```

Comment:
Should these annotations be set on MachineSet or MachineDeployment rather than the infrastructure template reference?

Reply:
yes, these are only for MachineSets and MachineDeployments, not infrastructure templates. i thought i had made it clear, but should i add more language to specify they are not for infrastructure templates?

Reply:
I was wondering if it'd actually be more appropriate to have them on infrastructure template resources rather than MS/MD

Reply:
i think my preference would be to leave it on the MS/MD so that users could override with a finer granularity (eg same template for 2 different MS/MD). we talked about this a little in the past, but i don't think we discussed moving the annotations to the infrastructure templates.